Release blog post for v1.8 (#648)
qdrant/landing_page: qdrant-landing/content/blog/qdrant-1.8.x.md

We optimized our search to improve throughput by a factor of 16.

This is based on our test benchmarks in this [Sparse vectors benchmark](https://github.com/qdrant/sparse-vectors-benchmark) repository. For those benchmarks, we ran
the full query dataset againt a system with 8 CPUs.

<!-- Not sure if we need the details of how we created the test collection -->

Based on our tests, you can expect significant performance improvements. 
The performance that we see is shown on the following two-dimensional histogram:

![Histogram with increasing latency for higher query dimensions](/blog/qdrant-1.8.x/neurIPS_bench_example.png)

The colors within the scatter plot show the frequency of the results. The "red"
points show the highest frequency.

## Optimized CPU use

We continue to optimize our search. With dynamic CPU saturation we've increased
indexing speed, with no impact on search requests.

This version introduces an `optimizer_cpu_budget` parameter to control the maximum number of CPUs used
for indexing. For most users, the default `optimizer_cpu_budget` works well. It allocates more CPU resources
for building indexes faster, while reserving resources to handle searches.

> 

Author: ```suggestion

![Allocation of CPU resources](/blog/qdrant-1.8.x/cpu-management.png)

```

With our [Collections](/documentation/concepts/collections/) API, you can 
configure how Qdrant saturates the CPUs in your configuration. 

## Optimize RAM with immutable text fields

We have optimized the required RAM with immutable text fields. We have set up
a field index which helps minimize what is stored.

Based on our tests of a system with 64GB of RAM, we've reduced by the amount of 
required RAM by around 10%. We expect greater improvements on systems with less RAM.

Mutable documents require additional RAM.

## Increase search performance

To improve search performance we have optimized the way we load documents for searches with a text field index. This also minimizes the load on RAM. 

To set this up, we load documents mostly sequentially, in increasing order.

## Minor improvements and new features

Beyond these enhancements, Qdrant v1.8.0 adds and improves several features:

<!-- Requires merging https://github.com/qdrant/landing_page/pull/631 -->
1. Order results by payload value with the Scroll API. For more information,
   see how you can [Order points by a payload key](/documentation/concepts/points/#order-points-by-a-payload-key).

> 1. [Implement datetime support for the payload index](https://github.com/qdrant/qdrant/issues/3320)

Reviewer: Maybe you want to link to the documentation page (after merging #678) instead of linking the issue.

Author: When we release v1.8 (and as @xzfc suggests, after merging #678) 

```suggestion
1. [Implement datetime support for the payload index](/documentation/concepts/filtering/#datetime-range)
```

---

Today, we are pleased to announce the release of [Qdrant 1.8.0](https://github.com/qdrant/qdrant/releases/tag/v1.8.0).
We've optimized performance in a number of ways. We've addressed issues with:

<!-- Shard deltas (deferred to v1.9) -->
- Sparse vectors
- Dynamic CPU saturation
- Text immutability
- Text loading optimization

We've also added <!-- TBD -->.

Did we miss something? We rely on your feedback to drive our development. We
welcome your contributions, especially in our [Discord community](https://qdrant.to/discord). Join us, introduce yourself, and help us build the best vector search engine!

## Improved sparse vector performance

The primary focus of Qdrant is performance. It's why we build our work in Rust.
We also want to optimze performance in "real-world" situations.

We used the [NeurIPS 2023 datasets](https://big-ann-benchmarks.com/neurips23.html). We've optimized our search to improve throughput by a factor of 16.

We set up test benchmarks in our [Sparse vectors benchmark](https://github.com/qdrant/sparse-vectors-benchmark) repository. For our benchmarks, we use
moderately-sized Azure instances with separate clients and servers:

> <!-- Arnaud isn't sure we should include these details. I think they're useful for our readers. (I've removed the Azure configuration bits described elsewhere.)  -->

Reviewer 2: I slept on this and here is what I prefer.

We want the users to know that the sparse vector search got faster and that we got there by using a standardized dataset (using a link to our benchmark code).

In the spirit of conciseness, we could share the current graph by saying that it is an example of running the full query dataset against a 8 CPU machine.

Author: Thanks @agourlay . You've simplified things. I hope [these changes](https://github.com/qdrant/landing_page/pull/648/files#diff-c37269758c74910bcc8e4dae1c4ccc2fb994517b6ef01c4a5e2aa55109daaf27R44-R55) lays out what you're looking for.

| System | vCPU | RAM (GB) |
|--------|------|----------|
| Server | 8    | 32       |
| Client | 4    | 16       |

We created the test collection with:

- 8 Sparse NeruIPS segments
- `index.on_disk=false`

Based on those datasets, we found the following two-dimentional histogram of
latency per query dimension count:

![Histogram with increasing latency for higher query dimensions](/blog/qdrant-1.8.x/neurIPS_bench_example.png)

The colors within the scatter plot show the frequency of the results. The "red"
points show the highest frequency.

> While this is a two-dimensional graph, the dataset represents over 100 dimensions! 

Reviewer 2: 100 dimensions is not super impressive, maybe we can remove this part :)

Author: Deleted.

## Optimized CPU use

> We continue to optimize our search to minimize the load on your hardware. One

Reviewer 3: Load is not minimized, but exactly the same.

But we plan when an how much CPU cores to use in a better way, mainly to index (a lot?) faster.

Author: Thank you for explaining. I've removed references to "load," and noted "increased indexing speed"

part of that is on CPUs. On a typical system, the loads on each CPU is 
low, which is a waste of resources.

We optimize how we use CPUs. With dynamic CPU saturation, we set an 
`optimizer_cpu_budget` to drive the number of CPU _cores_ to saturate with
optimization tasks. Specifically, if the value is:

- `0`: Qdrant keeps one or more CPU cores unallocated (default)
- A negative number: Qdrant subtracts this from the number of available CPU cores

> - A positive number: Qdrant assigns this exact number of CPU cores to your configuration

Reviewer 3: Adding these details here may be more confusing than helpful.

The general idea is that the user does not have to configure anything (but the posibility is there).

Something like this, but better worded, is probably better:

> With dynamic CPU saturation we find the right balance on your hardware to index faster while handling all search requests as normal.

Author: Thanks. I hope this [updated intro](https://github.com/qdrant/landing_page/pull/648/files#diff-c37269758c74910bcc8e4dae1c4ccc2fb994517b6ef01c4a5e2aa55109daaf27R59-R60) works.

Reviewer 3: It is not that we don't impact search at all. But we try to keep search performance at a reasonable level. That's why it's a 'good balance between indexing speed and search performance'.

> <!-- Question: where do we set the `optimizer_cpu_budget, and how does that relate to `max_indexing_threads` -->

Reviewer 3: In the configuration (`config.yaml`). There it also explains what the possible values are. It is very likely that users keep it at 0, which means: automatic.

See: https://github.com/qdrant/qdrant/blob/61fa4e74e4de9f08393a77a14489983a32700c85/config/config.yaml#L48-L52

Around it, there are other parameters that affect CPU utilization as well. But those have existed before.

Author: Based on your other comments, I'm guessing you're just answering my question. While I still reference the `optimizer_cpu_budget`, I've removed details. Users can search for that parameter in our code.

Reviewer 3: > Users can search for that parameter in our code.

In our configuration file, yes!

For most users, the default works well. It allocates most CPUs for building
indexes, while reserving CPUs to handle searches.

With our [Collections](/documentation/concepts/collections/) API, you can 
configure how Qdrant saturates the CPUs in your configuration. 

As shown in our API documenmtation, `max_indexing_threads` is a part of the
`hnsw_config` parameter. For more information see our 
[Create collection](ihttps://qdrant.github.io/qdrant/redoc/index.html#tag/collections/operation/create_collection) REST call.

The `max_indexing_threads` is the number of parallel threads used by Qdrant
to build your index in the background. The options are:

- `null`: No limit. Dynamically saturates your CPUs
- `0`: Automatically selects between 8 and 16 CPUs, to minimize the risk of

>   broken or inefficient HNSW graphs.

Reviewer 3: We already had this, and its probably more confusing than helpful. We've mostly kept it for backward compatibility.

I'd recommend not to cover this.

Author: Deleted references / discussion of `max_indexing_threads`.

If you have more CPUs, you could try different configurations. For example:

- Assume you have 128 CPU cores. You have also set up 4 Qdrant instances. If
you set `max_indexing_threads` to 32, you've allocated 25% of available CPU
cores for indexing and optimization. 
- If you need to reserve CPUs for important ongoing searches, you can set
`max_indexing_threads` to a negative value. Qdrant then reserves those cores

> for searches.

Reviewer 3: No. For this you change `optimizer_cpu_budget` and `max_indexing_threads` is unchanged. `optimizer_cpu_budget` is global per instance, the other is not.

I gave this example, but it's probably not the best example anyway as there are different ways to configure limits externally which are more reliable.

This parameter becomes interesting if a user wants to fine tune the balance between indexing and searching. But that is a very tricky topic, and we should not recommend it to 99% of the users.

Author: Since a post like this should not cover "edge cases," I've removed these bullets.

preview_image: /blog/qdrant-1.8.x/qdrant-1.8.0.png
title_preview_image: /blog/qdrant-1.8.x/qdrant-1.8.0.png
small_preview_image: /blog/qdrant-1.8.x/qdrant-1.8.0.png
date: 2024-02-26T11:12:00-08:00
author: Mike Jang
featured: false 
tags:
  - vector search
  - new features
  - sparse vector performance
  - dymamic CPU saturation
  - text immutability
  - text loading optimization
weight: 0 # Change this weight to change order of posts
---

Today, we are pleased to announce the release of [Qdrant 1.8.0](https://github.com/qdrant/qdrant/releases/tag/v1.8.0).
We've optimized performance in a number of ways. We've addressed issues with:

<!-- Shard deltas (deferred to v1.9) -->
- Sparse vectors
- Dynamic CPU saturation
- Text immutability
- Text loading optimization

We've also added <!-- TBD -->.

Did we miss something? We rely on your feedback to drive our development. We
welcome your contributions, especially in our [Discord community](https://qdrant.to/discord). Join us, introduce yourself, and help us build the best vector search engine!

## Improved sparse vector performance

The primary focus of Qdrant is performance. It's why we build our work in Rust.

> We also want to optimze performance in "real-world" situations.

Reviewer 2: ```suggestion
We also want to optimize performance in "real-world" situations.
```

> We used the [NeurIPS 2023 datasets](https://big-ann-benchmarks.com/neurips23.html). We've optimized our search to improve throughput by a factor of 16.

Reviewer 2: The style is a bit heavy here IMHO, having three short sentences starting with `we`.

Author: I actually had 4 in a row.... :laughing:

> We used the [NeurIPS 2023 datasets](https://big-ann-benchmarks.com/neurips23.html). We've optimized our search to improve throughput by a factor of 16.

Reviewer 2: maybe we could add more context

```suggestion
We used the [NeurIPS 2023 datasets](https://big-ann-benchmarks.com/neurips23.html) which are a standard to evaluate vector search algorithms.
```

We set up test benchmarks in our [Sparse vectors benchmark](https://github.com/qdrant/sparse-vectors-benchmark) repository. For our benchmarks, we ran
the full query dataset againt a system with 8 CPUs.

<!-- Not sure if we need the details of how we created the test collection -->

> Based on our tests, we found the following two-dimentional histogram of

Reviewer 2: I'd like to make it more explicit that the graph reflects the new performance that users can expect.

Author: Hopefully... https://github.com/qdrant/landing_page/pull/648/files#r1507832960

latency per query dimension count:

![Histogram with increasing latency for higher query dimensions](/blog/qdrant-1.8.x/neurIPS_bench_example.png)

The colors within the scatter plot show the frequency of the results. The "red"
points show the highest frequency.

## Optimized CPU use

We continue to optimize our search. With dynamic CPU saturation we've increased
indexing speed, with no impact on search requests.

For most users, the default `optimizer_cpu_budget` works well. It allocates
more CPU resources for building indexes, while reserving resources to handle

> searches.

Reviewer 3: If talking about this parameter flatout like this, it may be a bit confusing. It is a new thing.

Maybe we can word it something like this:

> This version introduces a `optimizer_cpu_budget` parameter to control the maximum number of CPUs used for indexing. For most users, the default `optimizer_cpu_budget` works well. It allocates
more CPU resources for building indexes faster, while reserving resources to handle
searches.

Author: ```suggestion
This version introduces an `optimizer_cpu_budget` parameter to control the maximum number of CPUs used
for indexing. For most users, the default `optimizer_cpu_budget` works well. It allocates more CPU resources
for building indexes faster, while reserving resources to handle searches.
```

preview_image: /blog/qdrant-1.8.x/qdrant-1.8.0.png
title_preview_image: /blog/qdrant-1.8.x/qdrant-1.8.0.png
small_preview_image: /blog/qdrant-1.8.x/qdrant-1.8.0.png
date: 2024-02-26T11:12:00-08:00
author: Mike Jang
featured: false 
tags:
  - vector search
  - new features
  - sparse vector performance
  - dymamic CPU saturation
  - text immutability
  - text loading optimization
weight: 0 # Change this weight to change order of posts
---

Today, we are pleased to announce the release of [Qdrant 1.8.0](https://github.com/qdrant/qdrant/releases/tag/v1.8.0).
We've optimized performance in a number of ways. We've addressed issues with:

<!-- Shard deltas (deferred to v1.9) -->
- Sparse vectors
- Dynamic CPU saturation
- Text immutability
- Text loading optimization

We've also added <!-- TBD -->.

Did we miss something? We rely on your feedback to drive our development. We
welcome your contributions, especially in our [Discord community](https://qdrant.to/discord). Join us, introduce yourself, and help us build the best vector search engine!

## Improved sparse vector performance

The primary focus of Qdrant is performance. It's why we build our work in Rust.

> We also want to optimize performance in "real-world" situations.

Author: ```suggestion
It helps us optimize performance in "real-world" situations.
```

> 
> We used the [NeurIPS 2023 datasets](https://big-ann-benchmarks.com/neurips23.html) which are a standard to evaluate vector search algorithms.

Author: Building from the 1.7 release

```suggestion

For our v1.7.0 release we introduced sparse vectors. As noted in our [Qdrant v.1.7.0 release announcement](https://qdrant.tech/articles/qdrant-1.7.x/#sparse-vectors), we store sparse vectors differently and keep the non-zero dimensions.

As an additional improvement, we used the [NeurIPS 2023 datasets](https://big-ann-benchmarks.com/neurips23.html) which are a standard to evaluate vector search algorithms.
```

We optimized our search to improve throughput by a factor of 16.

> We set up test benchmarks in our [Sparse vectors benchmark](https://github.com/qdrant/sparse-vectors-benchmark) repository. For our benchmarks, we ran

Author: ```suggestion
This is based on our test benchmarks in this [Sparse vectors benchmark](https://github.com/qdrant/sparse-vectors-benchmark) repository. For those benchmarks, we ran
```

the full query dataset againt a system with 8 CPUs.

<!-- Not sure if we need the details of how we created the test collection -->

> Based on our tests, we found the following two-dimentional histogram of
> latency per query dimension count:

Author: ```suggestion
Based on our tests, you can expect significant performance improvements. 
The performance that we see is shown on the following two-dimensional histogram:
```
